src/brogue/Items.c

```diff
@@ -5511,16 +5489,164 @@ boolean chooseTarget(short returnLoc[2],
+    }
+
+    switch (magicPolarity) {
+        case MAGIC_POLARITY_MALEVOLENT:
+            if (goodKinds == 0) {
+                kindCount = 0;
+            } else {
+                kindCount = totalKinds - goodKinds;
+            }
+            break;
+        case MAGIC_POLARITY_BENEVOLENT:
+            kindCount = goodKinds;
+            break;
+        default:
+            kindCount = totalKinds;
+            break;
+    }
+    return kindCount;
+}
+
+// Gets the final unidentified item kind for the given category and magic polarity, if any. Returns -1 otherwise.
+// Pass magicPolarity = 0 to get the last item kind for the category, ignoring magic polarity
+int tryGetLastUnidentifiedItemKind(enum itemCategory category, int magicPolarity) {
+    int lastItemKind = -1;
+    int totalItemKinds = itemKindCount(category, 0);
+    itemTable *theItemTable = tableForItemCategory(category);
+
+    if (theItemTable && (totalItemKinds > 0)) {
+        for (int i = 0; i < totalItemKinds; i++) {
+            if (!(theItemTable[i].identified)
+                && (theItemTable[i].magicPolarity == magicPolarity || magicPolarity == MAGIC_POLARITY_ANY)) {
+                if (lastItemKind != -1) {
+                    return -1; // At least two unidentified items remain.
+                }
+                lastItemKind = i;
+            }
+        }
+    }
+    return lastItemKind;
+}
+
+// Counts the number of items where the magic polarity has been revealed for the given item category and magic polarity
+int magicPolarityRevealedItemKindCount(enum itemCategory category, int magicPolarity) {
+    int kindCount = -1;
+    int totalItemKinds = itemKindCount(category, 0);
+    itemTable *theItemTable = tableForItemCategory(category);
+
+    if (theItemTable && (totalItemKinds > 0) && magicPolarity) {
+        kindCount = 0;
+        for (int i = 0; i < totalItemKinds; i++) {
+            if (theItemTable[i].magicPolarity == magicPolarity &&
+                    (theItemTable[i].identified || theItemTable[i].magicPolarityRevealed)) {
+                kindCount += 1;
+            }
+        }
+    }
+    return kindCount;
+}
+
+// We can identify the last unidentified item kind...
+// 1. Of a category (for regression)
+// 2. Of a given polarity within a category if...
+// A. Its polarity is known
+// B. All items of the opposite polarity are either identified or their polarity is known
+void tryIdentifyLastItemKind(enum itemCategory category, int magicPolarity) {
+    int oppositeMagicPolarity = magicPolarity * -1;
+    itemTable *theItemTable = tableForItemCategory(category);
+    int lastItemKind = tryGetLastUnidentifiedItemKind(category, magicPolarity);
+    int oppositeCount, oppositeRevealedCount;
+
+    if (lastItemKind >= 0) {
+        oppositeRevealedCount = magicPolarityRevealedItemKindCount(category, oppositeMagicPolarity);
+        oppositeCount = itemKindCount(category, oppositeMagicPolarity);
+        if (magicPolarity == MAGIC_POLARITY_ANY
```

Comment:
I think putting the opposite stuff in an else from the ==ANY test will help readability here

Reply:
Done